### PR TITLE
fix: codeInput implement specs, code improvement and new readme examples

### PR DESCRIPTION
**Pull request by blakmetall**
fix: codeInput accessibility, a few code improvements and readme examples added

### CODE REVIEW

File: src/components/CodeInput/index.js [hunk 1]
```diff
@@ -82,6 +83,8 @@ const CodeInput = React.forwardRef((props, ref) => {
                     <RequiredAsterisk required={required} />
                     {label}
                 </StyledLabel>
+
+                <StyledHiddenLabel id={labelId}>{label}</StyledHiddenLabel>
```
Comment (LeandroTorresSicilia) [inline src/components/CodeInput/index.js:87]: Take a look here: https://www.w3.org/WAI/tutorials/forms/custom-controls/
In order to have the right a11y we should have an label for each input but since we don't have a clear understanding of how a11y should be implemented here we should not do this for now

File: src/components/CodeInput/inputItems.js [hunk 2]
```diff
@@ -46,6 +48,7 @@ const InputItems = React.forwardRef((props, ref) => {
                 ref={inputRef}
                 id={inputId}
                 isActive={isActive}
+                ariaLabelledBy={ariaLabelledBy}
```
Comment (LeandroTorresSicilia) [inline src/components/CodeInput/inputItems.js:51]: same ^^ let's not do this for now

File: src/components/CodeInput/inputItem.js [hunk 3]
```diff
@@ -54,6 +51,8 @@ const InputItem = React.forwardRef((props, ref) => {
             maxLength="1"
             id={id}
             isActive={isActive}
+            aria-label="Enter a number"
+            aria-labelledby={ariaLabelledBy}
```
Comment (LeandroTorresSicilia) [inline src/components/CodeInput/inputItem.js:55]: same ^^

File: src/components/CodeInput/index.js (hunk 1 above)
Comment (blakmetall) [reply src/components/CodeInput/index.js:87]: sure, definitely, I applied changes from my a11y perspective, might be diffuse, I understand completely... however it was quite a new learning and good exercise.

File: src/components/CodeInput/inputItems.js (hunk 2 above)
Comment (blakmetall) [reply src/components/CodeInput/inputItems.js:51]: roger that

File: src/components/CodeInput/inputItem.js (hunk 3 above)
Comment (blakmetall) [reply src/components/CodeInput/inputItem.js:55]: roger that again haha

File: integration/specs/CodeInput/codeInput-1.spec.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    beforeAll(() => {
+        browser.url('/#!/CodeInput/1');
+    });
+    beforeEach(() => {
+        browser.refresh();
+        const component = $(CODEINPUT);
+        component.waitForExist();
+    });
+    /*
+    it('should have focus on the first input on load', () => {
```
Comment (blakmetall) [inline integration/specs/CodeInput/codeInput-1.spec.js:15]: fixed: uncommented code in next commit

File: integration/specs/CodeInput/codeInput-1.spec.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        const component = $(CODEINPUT);
+        component.waitForExist();
+    });
+    it('should have focus on the first input on load', () => {
+        const codeInput = new PageCodeInput(CODEINPUT);
+        expect(codeInput.getFocusedIndex()).toBe(0);
+    });
+    it('should change focus to the second input when the first input its filled with a number', () => {
+        const codeInput = new PageCodeInput(CODEINPUT);
+        browser.keys(['', '2']);
```
Comment (LeandroTorresSicilia) [inline integration/specs/CodeInput/codeInput-1.spec.js:21]: why we need to pass ```['', '2']```?

File: integration/specs/CodeInput/codeInput-1.spec.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        expect(codeInput.getFocusedIndex()).toBe(1);
+    });
+    it('should remove value from first input when first input is selected, has a value and "Backspace" key is pressed', () => {
+        const codeInput = new PageCodeInput(CODEINPUT);
+        browser.keys(['', '2']);
+        expect(codeInput.getInputValueByIndex(0)).toBe('2');
```
Comment (LeandroTorresSicilia) [inline integration/specs/CodeInput/codeInput-1.spec.js:27]: ```getInputValueAtIndex```

File: integration/specs/CodeInput/codeInput-1.spec.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        browser.keys(DELETE_KEY);
+        expect(codeInput.getInputValueByIndex(0)).toBe('');
+    });
+    it('should return focus to the first input and clear his value when the first input has a value, the second input is empty, is focused and "Backspace" key is pressed', () => {
+        const codeInput = new PageCodeInput(CODEINPUT);
+        browser.keys(['', '2']);
+        expect(codeInput.getInputValueByIndex(0)).toBe('2');
+        expect(codeInput.getFocusedIndex()).toBe(1);
+        browser.keys(DELETE_KEY);
+        expect(codeInput.getInputValueByIndex(0)).toBe('');
```
Comment (LeandroTorresSicilia) [inline integration/specs/CodeInput/codeInput-1.spec.js:37]: in the test name say `should return focus to the first input`, then we should add a expect here:
```expect(codeInput.getFocusedIndex()).toBe(0);```

File: integration/specs/CodeInput/codeInput-1.spec.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        browser.keys(['', '2']);
+        browser.keys(['', '2']);
+        browser.keys(['', '2']);
+        expect(codeInput.getFocusedIndex()).toBe(3);
+    });
+    it('should keep focus in the current input when we type a letter', () => {
+        const codeInput = new PageCodeInput(CODEINPUT);
+        expect(codeInput.getFocusedIndex()).toBe(0);
+        browser.keys(['', 'A']);
+        expect(codeInput.getFocusedIndex()).toBe(0);
```
Comment (LeandroTorresSicilia) [inline integration/specs/CodeInput/codeInput-1.spec.js:58]: add another test like this that check for the input to be empty in order to prove that it does not accepts other values than numbers

File: integration/specs/CodeInput/codeInput-1.spec.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    });
+    it('should keep current input focus when unfocused inputs are clicked', () => {
+        const codeInput = new PageCodeInput(CODEINPUT);
+        expect(codeInput.getFocusedIndex()).toBe(0);
+        codeInput.clickInputByIndex(2);
```
Comment (LeandroTorresSicilia) [inline integration/specs/CodeInput/codeInput-1.spec.js:63]: clickInputAtIndex

File: integration/specs/CodeInput/codeInput-1.spec.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        browser.keys(['', '2']);
+        expect(codeInput.getFocusedIndex()).toBe(1);
+        codeInput.clickInputByIndex(0);
+        expect(codeInput.getFocusedIndex()).toBe(1);
+    });
+    it('should leave focus on all inputs when "Tab" key is pressed', () => {
+        const codeInput = new PageCodeInput(CODEINPUT);
+        expect(codeInput.getFocusedIndex()).toBe(0);
+        browser.keys(TAB_KEY);
+        expect(codeInput.getFocusedIndex()).toBe(-1);
```
Comment (LeandroTorresSicilia) [inline integration/specs/CodeInput/codeInput-1.spec.js:77]: let's use undefined to indicate that no input is focused

File: src/components/CodeInput/__test__/codeInput.spec.js [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    });
+    it('should call useValueState with the right value and length', () => {
+        useValueState.mockImplementation(() => ['1', '', '', '']);
+        mount(<CodeInput value="1" />);
+        expect(useValueState).toHaveBeenCalledWith('1', 4);
+    });
+    it('should call useFocusedIndexState with the right value and length', () => {
+        useFocusedIndexState.mockImplementation(() => 1);
+        mount(<CodeInput value="1" />);
+        expect(useValueState).toHaveBeenCalledWith('1', 4);
```
Comment (LeandroTorresSicilia) [inline src/components/CodeInput/__test__/codeInput.spec.js:146]: when testing with what was called a mocked function you need to call mockReset before since in the last test useValueState  was called with this same value then this is not testing really that now it was called with ```'1', 4```

File: integration/specs/CodeInput/codeInput-1.spec.js (hunk 5 above)
Comment (blakmetall) [reply integration/specs/CodeInput/codeInput-1.spec.js:25]: My bad: I though it was necessary the first value, it refers to Metadata, I changed to key only and seems to be working...

- - - - -

Reverted to browsers.keys('2'); all of them in next PR update

-

File: src/components/CodeInput/__test__/inputItems.spec.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+                    .prop('value'),
+            ).toBe(value);
+        });
+    });
+    it('should have unique ids on each input identified by base id and input index', () => {
+        const values = ['', '', '', ''];
+        const id = 'code-input-base';
+        const component = mount(<InputItems value={values} id={id} />);
+        values.forEach((value, index) => {
+            const singleId = `${id}-${index}`;
```
Comment (LeandroTorresSicilia) [inline src/components/CodeInput/__test__/inputItems.spec.js:49]: we should construct this id using our uniqueId hook instead of use the index since it will assure that the id will be unique, remember that id can't be twice in the entire web page

File: src/components/CodeInput/__test__/inputItems.spec.js [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+        const input0 = component
+            .find(InputItem)
+            .at(0)
+            .getDOMNode();
+        const input1 = component
+            .find(InputItem)
+            .at(1)
+            .getDOMNode();
+        expect(input0).not.toBe(ref.current);
+        expect(input1).toBe(ref.current);
```
Comment (LeandroTorresSicilia) [inline src/components/CodeInput/__test__/inputItems.spec.js:72]: to compare object values you must use toEqual

File: integration/specs/CodeInput/codeInput-1.spec.js (hunk 10 above)
Comment (blakmetall) [reply integration/specs/CodeInput/codeInput-1.spec.js:93]: I have 3 methods returning -1, 

should I apply undefined return value on all of them or just for gestFocusedIndex ?

File: integration/specs/CodeInput/codeInput-1.spec.js (hunk 7 above)
Comment (blakmetall) [reply integration/specs/CodeInput/codeInput-1.spec.js:46]: definitely... that was the main test for that block

File: src/components/CodeInput/hooks/__test__/usePreviousIndex.spec.js [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+        const { result } = renderHook(() => usePreviousIndex());
+        expect(result.current).toBe(undefined);
+    });
+    it('should return 1 when focusedIndex is 1', () => {
+        const { result, rerender } = renderHook(initialValue => usePreviousIndex(initialValue));
+        expect(result.current).toBe(undefined);
+        rerender(2);
+        setTimeout(() => {
+            expect(result.current).toEqual(2);
+        }, 1000);
```
Comment (LeandroTorresSicilia) [inline src/components/CodeInput/hooks/__test__/usePreviousIndex.spec.js:15]: why use setTimeout? if rerender is async then use async await in order to wait for the result

File: src/components/CodeInput/index.js [hunk 15]
```diff
@@ -113,8 +116,6 @@ const CodeInput = React.forwardRef((props, ref) => {
 CodeInput.propTypes = {
     /** The id of the outer element. */
     id: PropTypes.string,
-    /** An identifier for the CodeInput element. */
-    name: PropTypes.string,
```
Comment (LeandroTorresSicilia) [inline src/components/CodeInput/index.js:117]: remember to remove name prop in index.d.ts

File: src/components/CodeInput/inputItems.js [hunk 16]
```diff
@@ -20,11 +21,19 @@ const InputItems = React.forwardRef((props, ref) => {
...
+            onFocus(e);
+        }
+    };
+
     const inputs = value.map((inputValue, index) => {
         const inputIndex = index;
         const isActive = inputIndex === focusedIndex;
         const inputRef = isActive ? ref : null;
-        const inputId = isActive ? id : null;
+        const inputId = `${id}-${index}`;
```
Comment (LeandroTorresSicilia) [inline src/components/CodeInput/inputItems.js:36]: as said before input id should be generated with uniqueId, but since we will not associate it with no label then it is not needed then, for now, let's not pass id to the input

File: src/components/CodeInput/inputItems.js [hunk 17]
```diff
@@ -20,11 +21,19 @@ const InputItems = React.forwardRef((props, ref) => {
         id,
     } = props;
 
+    const handleOnFocus = e => {
+        if (readOnly || disabled) {
+            ref.current.blur();
```
Comment (LeandroTorresSicilia) [inline src/components/CodeInput/inputItems.js:26]: check this, because I think that when you pass disabled or readonly to the input itself then it does not get the focus so you don't need to call blur programatically

File: src/components/CodeInput/__test__/inputItems.spec.js (hunk 12 above)
Comment (blakmetall) [reply src/components/CodeInput/__test__/inputItems.spec.js:49]: For this case, we are using the uniqueID hook in the parent

`const inputId = useUniqueIdentifier('code-input');`

Then it is passed to InputItems which renders unique ids for inputs based on parentID and index, setting there a unique id for each input.

For my that solution work, if not for learning purposes if possible can you share why?

--

File: src/components/CodeInput/pageObject/index.js [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    getFocusedInput() {
+        const inputs = $(this.rootElement).$$('input');
+        return inputs.filter(input => input.isFocused()).shift();
+    }
+
+    getInputByIndex(inputIndex) {
+        const inputs = $(this.rootElement).$$('input');
+        return inputs.filter((elem, index) => index === inputIndex).shift();
```
Comment (LeandroTorresSicilia) [inline src/components/CodeInput/pageObject/index.js:83]: I think you can take access the input by index directly, look at other implementations, thus we avoid the filter

File: src/components/CodeInput/styled/index.js [hunk 19]
```diff
@@ -83,6 +89,11 @@ const StyledInput = attachThemeAttrs(styled.input)`
     `}
 `;
 
+const StyledInputItems = attachThemeAttrs(styled.div)`
+    position: relative;
+    margin-bottom: -12px;
```
Comment (LeandroTorresSicilia) [inline src/components/CodeInput/styled/index.js:94]: @TahimiLeonBravo check this negative margin

File: integration/specs/CodeInput/codeInput-1.spec.js [hunk 20]
```diff
@@ -0,0 +1,120 @@
+const PageCodeInput = require('../../../src/components/CodeInput/pageObject');
+const { DELETE_KEY, TAB_KEY } = require('../../constants');
+
+const CODEINPUT = '#codeinput-1';
+
+const awaitForUpdate = () => {
+    browser.pause(99);
+};
```
Comment (LeandroTorresSicilia) [inline integration/specs/CodeInput/codeInput-1.spec.js:8]: we should avoid use pause, it is not deterministic

File: src/components/CodeInput/__test__/inputItems.spec.js (hunk 12 above)
Comment (blakmetall) [reply src/components/CodeInput/__test__/inputItems.spec.js:49]: test removed for now because we are removing id's over the inputs for now; until we fully integrate accesibility.

File: src/components/CodeInput/styled/index.js (hunk 19 above)
Comment (blakmetall) [reply src/components/CodeInput/styled/index.js:94]: Screen Shot 2020-05-14 at 16 56 14

that was because in a previous commit I added to inputs a 12px margin-bottom to prepare for responsiveness,  so I added to the container a negative margin bottom to bring back moved the position.

Screen Shot 2020-05-14 at 16 57 47

File: src/components/CodeInput/inputItems.js (hunk 17 above)
Comment (blakmetall) [reply src/components/CodeInput/inputItems.js:26]: I believe that when I did this test my used example was wrong, that make me think I should disable programmatically... this is misterious hahaha...

Now I confirmed it is not needed

File: integration/specs/CodeInput/codeInput-1.spec.js [hunk 21]
```diff
@@ -80,8 +72,7 @@ describe('CodeInput base example', () => {
     });
     it('should keep focus on the second input when any input is clicked and the first input has a number already', () => {
         const codeInput = new PageCodeInput(CODEINPUT);
-        browser.keys(['', '2']);
-        awaitForUpdate();
+        browser.keys('2');
         expect(codeInput.getFocusedIndex()).toBe(1);
         codeInput.clickInputByIndex(0);
```
Comment (LeandroTorresSicilia) [inline integration/specs/CodeInput/codeInput-1.spec.js:77]: clickInputByIndex  => clickInputAtIndex

**Comment by TahimiLeonBravo** [comment]
@blakmetall  I added a design for readOnly appearance, let's follow these styles
https://zpl.io/bPjBNym
Also, add the value to the example when it is readOnly

File: src/components/CodeInput/readme.md [hunk 22]
```diff
@@ -88,6 +88,109 @@ function CodeInputBase() {
 };
 
 <div className="rainbow-align-content_center rainbow-m-around_xx-large">
-    <CodeInputBase  />
+    <CodeInputWithLength  />
+</div>
+```
+
+##### CodeInput events:
```
Comment (TahimiLeonBravo) [inline src/components/CodeInput/readme.md:95]: I think we don't need this example

File: src/components/CodeInput/styled/index.js (hunk 19 above)
Comment (TahimiLeonBravo) [reply src/components/CodeInput/styled/index.js:94]: @blakmetall @LeandroTorresSicilia I think we should not add this negative margin since this a bad practice if we want to separate the bottomHelpText or the errorText from the input we should remove the `margin-top`

File: src/components/CodeInput/inputItems.js [hunk 23]
```diff
@@ -44,18 +43,16 @@ const InputItems = React.forwardRef((props, ref) => {
...
                 onPaste={onPaste}
                 ref={inputRef}
-                id={inputId}
                 isActive={isActive}
             />
         );
     });
 
-    return <div>{inputs}</div>;
+    return <StyledInputItems>{inputs}</StyledInputItems>;
```
Comment (TahimiLeonBravo) [inline src/components/CodeInput/inputItems.js:51]: I think we should not need this wrapper here, instead, we should use a `Fragment`

**Comment by TahimiLeonBravo** [comment]
@blakmetall we need fix some tests here in order to merge this PR

File: src/components/CodeInput/pageObject/index.js [hunk 24]
```diff
@@ -12,6 +12,13 @@ class PageCodeInput {
         this.rootElement = rootElement;
     }
 
+    typeKey(key) {
```
Comment (LeandroTorresSicilia) [inline src/components/CodeInput/pageObject/index.js:15]: let's use type

**Comment by blakmetall** [comment]
- [x] type method from pageObject updated
- [x] spec removed for readOnly verification (not needed anymore)
- [x] readOnly styled configuration update
- [x] update over onClick event
- [x] removal of unnecessary condition related to readOnly (not needed anymore)

**Comment by blakmetall** [comment]
Note: integration tests failed over another Component unrelated tu CodeInput

Screen Shot 2020-05-21 at 14 10 28
